### ISSUE TITLE
How to enable C++ API usage from AAR builds (and integrate into gradle project)?

### ISSUE THREAD

**Issue by orangesomethingorange** [tensorflow/tensorflow#39148]
**System information**
- OS Platform and Distribution: Linux Ubuntu 16.04
- Mobile device: Samsung Galaxy S10
- TensorFlow installed from: pip
- TensorFlow version: 2.1.0
- Python version: 3.5.2
- Installed using: virtualenv / pip
- Bazel version (if compiling from source): 3.0.0
- GCC/Compiler version (if compiling from source): 5.4.0

**Describe the problem**

How to enable C++ API usage from AAR builds (and integrate into gradle project)?
I'm trying to run a simple inference from the C++ API... what am I missing? I've tried following the guides at the TF website, and the Github repositories (minimal / image classification) but I'm missing the part where the TFLite library joins the Android Studio NDK.

Is there a way to use C++ API the same way I've been using the Java API via gradle? (kindly see attached comment by @jdduke)

**Provide the exact sequence of commands / steps that you executed before running into the problem**

https://www.tensorflow.org/lite/guide/android
https://www.tensorflow.org/lite/guide/inference
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/examples/minimal/minimal.cc
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/examples/label_image/label_image.cc
https://www.tensorflow.org/lite/guide/build_arm64

**Any other info / logs**
_Originally posted by @jdduke in https://github.com/tensorflow/tensorflow/issues/34489#issuecomment-558261882_

**Comment by Saduf2019** [comment]
@orangesomethingorange 
Please refer to these issues and let us know if it helps:
#39120
[link](https://stackoverflow.com/questions/35140809/android-aar-package-for-native-library)
[link2](https://github.com/tensorflow/tensorflow/issues/34489)
[link3](https://medium.com/@yushulx/how-to-build-so-library-files-into-aar-bundle-in-android-studio-a44387c9a012)

**Comment by pranv12** [comment]
@orangesomethingorange 
Please check this also 
[link](https://stackoverflow.com/questions/49834875/problems-with-using-tensorflow-lite-c-api-in-android-studio-project)

**Comment by jdduke** [comment]
Hi @orangesomethingorange, we currently don't have any plans to expose the C++ API directly from the .aar. Generally speaking, using C++ across shared library boundaries is a tricky affair.

That said, we do expose the *C API* for TFLite in the prebuilt .aars. We're working on a sample which downloads the aar from Maven and extracts them via Gradle. The .aar also include the C API headers. See also https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/c.

**Comment by orangesomethingorange** [comment]
> @orangesomethingorange
> Please refer to these issues and let us know if it helps:
> #39120
> [link](https://stackoverflow.com/questions/35140809/android-aar-package-for-native-library)
> [link2](https://github.com/tensorflow/tensorflow/issues/34489)
> [link3](https://medium.com/@yushulx/how-to-build-so-library-files-into-aar-bundle-in-android-studio-a44387c9a012)

Hi @Saduf2019,
Thanks for the links ;)

### Trying to add the flatbuffers libs:
To use tfilte I need to also add the flatbuffers libraries, if I understand correctly.
I tried following [flatbuffers build guide](https://google.github.io/flatbuffers/flatbuffers_guide_building.html) and compile the libs using Android Studio's CMake.
Indirect path resulted in an error, so I set the directory's full path:

**CMakeLists.txt**:
```
cmake_minimum_required(VERSION 3.4.1)
set( FLATBUFFERS_SRC_DIR /full/path/to/flatbuffers )
add_subdirectory(${FLATBUFFERS_SRC_DIR}
                 ${CMAKE_CURRENT_BINARY_DIR}/flatbuffers-build
                 EXCLUDE_FROM_ALL)
add_library(
             native-lib
             SHARED
             native-lib.cpp )
find_library(
              log-lib
              log )
target_link_libraries(
        native-lib
        PRIVATE flatbuffers
        ${log-lib} )
```

but trying to build this with Android Studio results in the following:

```
FAILED: flatbuffers-build/tests/union_vector/union_vector_generated.h 
g���d�C��: not found
FAILED: flatbuffers-build/tests/namespace_test/namespace_test2_generated.h 
g���d�C��: not found
FAILED: flatbuffers-build/tests/arrays_test_generated.h 
g���d�C��: not found
FAILED: flatbuffers-build/tests/native_type_test_generated.h 
g���d�C��: not found
...
FAILED: flatbuffers-build/tests/monster_test_bfbs_generated.h 
g���d�C��: not found
ninja: build stopped: subcommand failed.
```

### Trying to add the tflite libs:
I tried following [TF world's pdf](https://github.com/infil00p/tf_world) and [this question](https://stackoverflow.com/questions/49834875/problems-with-using-tensorflow-lite-c-api-in-android-studio-project) to use a .so library.
Also I tried using Bazel to build .a files or .aar files - and include them in Android Studio.
I've used Bazel on my server and also within TensorFlow's Docker container.
So after all these trials I'm left with the .so, .a and .aar files.
They're all currently located inside my Android app folder, under /app/libs directory, and I've tried to include them in my CMakeLists.txt file in several ways (not all at once, of course):

**CMakeLists.txt**:

```
# set(tflite_DIR ../../../libs)
# set(tflite_DIR /full/path/to/app/libs )
# set(TENSORFLOW_DIR /full/path/to/tensorflow/ )

#add_library(lib_tflite SHARED IMPORTED)
#set_target_properties(lib_tflite PROPERTIES IMPORTED_LOCATION
#        ${tflite_DIR}/tflite/lib/${ANDROID_ABI}/libtensorflowlite_c.so)
#
#add_library(lib_tflite_gpu SHARED IMPORTED)
#set_target_properties(lib_tflite_gpu PROPERTIES IMPORTED_LOCATION
#        ${tflite_DIR}/tflite/lib/${ANDROID_ABI}/libtensorflowlite_gpu_delegate.so)
#
#add_library(lib_nnapi SHARED IMPORTED)
#set_target_properties(lib_nnapi PROPERTIES IMPORTED_LOCATION
#        ${tflite_DIR}/tflite/lib/${ANDROID_ABI}/libnnapi_delegate.so)
#
#target_include_directories(hello-libs PRIVATE
#        ${tflite_DIR}/tflite/include
#        ${tflite_DIR}/tflite/include/tensorflow/lite/tools/make/downloads
#        ${tflite_DIR}/tflite/include/tensorflow/lite/tools/make/downloads/flatbuffers/include
#        )

#add_library(libtflite SHARED IMPORTED)
#
#set_target_properties(libtflite PROPERTIES IMPORTED_LOCATION
#        ${TENSORFLOW_DIR}bazel-out/arm64-v8a-opt/bin/tensorflow/lite/libtensorflowLite.so)

#add_library(
#        libtflite
#        SHARED
#        IMPORTED )
#
#set_target_properties(libtflite
#        PROPERTIES IMPORTED_LOCATION
#        ${tflite_DIR}/tensorflow-lite.aar)

#add_library(libtensorflowLite
#        SHARED
#        IMPORTED)
#
#set_target_properties(
#        libtensorflowLite PROPERTIES IMPORTED_LOCATION
#        ${tflite_DIR}/libtensorflowLite.so)
```

so for example here is something that builds and compiles without any errors:
```
set(tflite_DIR /full/path/to/app/libs )
add_library(libtensorflowLite
        SHARED
        IMPORTED)
set_target_properties(
        libtensorflowLite PROPERTIES IMPORTED_LOCATION
        ${tflite_DIR}/libtensorflowLite.so)
target_link_libraries(
        native-lib
        libtensorflowLite
        ${log-lib} )
```

but when I try to use it in my cpp code - it doesn't seem to recognize any tensorflow library.
I can't `#include` anything or use any functions, ops or headers.
Is there an extra step I'm missing here?

**Comment by orangesomethingorange** [comment]
> @orangesomethingorange
> Please check this also
> [link](https://stackoverflow.com/questions/49834875/problems-with-using-tensorflow-lite-c-api-in-android-studio-project)

thanks @pranv12,
I checked that link too - see my previous reply.

**Comment by orangesomethingorange** [comment]
> Hi @orangesomethingorange, we currently don't have any plans to expose the C++ API directly from the .aar. Generally speaking, using C++ across shared library boundaries is a tricky affair.
> 
> That said, we do expose the _C API_ for TFLite in the prebuilt .aars. We're working on a sample which downloads the aar from Maven and extracts them via Gradle. The .aar also include the C API headers. See also https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/c.

@jdduke that's great news - I'll check that out now :)
Generally speaking I'm currently trying to run a simple inference from a .tflite model file. When using the Java API I ran the inference and used the NNAPI and GPU delegates.
From the README description I can tell that "This includes C APIs for common types, like kernels and delegates, as well as an explicit C API for inference" so I guess that it can work.

Later on I'll want to do some pre- or post- processing to the inputs / outputs of the image - I assume that's where the C API differs from the C++ one - correct?

Perhaps including an external library - such as OpenCV - may solve this.
Any thoughts?

**Comment by pranv12** [comment]
@orangesomethingorange  

Reply for steps you did in [comment](https://github.com/tensorflow/tensorflow/issues/39148#issuecomment-624075335)
Whichever IDE you are using search there for the option to include additional dependencies. And then try adding path to TFlite folder. This will add your TFlite Libraries.
For flatbuffers please read below.


I had same problem. 
Here's what I did.

1. In my workspace, I created a folder `include` (for External dependencies eg. TFlite, flatbuffers, etc)
2. Dowloaded only `tensorflow` folder from [source](https://github.com/tensorflow/tensorflow/tree/master/tensorflow). (You can do it using [svn](https://tortoisesvn.net/docs/release/TortoiseSVN_en/index.html))
3. Similarly downloaded TFLite dependency [flatbuffers](https://github.com/google/flatbuffers/tree/master/include/flatbuffers) folder from the source. 
4. Also, TFlite needs absl header files so I downloaded [absl header files](https://github.com/abseil/abseil-cpp/tree/master/absl) from the source.
5. I'm working in Microsoft VS. I added the `path/to/dir/include` in additional dependencies. Whichever IDE you are using search there for the option to include additional dependencies.

Only the above folders are necessary no need to clone the complete repo.

After this include folder contains:
include:
--tensorflow
    ---- Some Folders (Including lite)
    ---- Some Files
--flatbuffers
   ---- Some Folders
   ---- Some Files
--absl
   ---- Some Folders
   ---- Some Files


Extra note: 
To download only specific folder from GitHub repository.
A tool called svn is helpful.
Installation
On Ubuntu/Linux:
`sudo apt install subversion`
On Windows:
[Download](https://tortoisesvn.net/downloads.html) the latest installer and install.

Usage: 
E.g. to download [tensorflow ](https://github.com/tensorflow/tensorflow/tree/master/tensorflow)folder from tensorflow repository
Navigate to folder:
`https://github.com/tensorflow/tensorflow/tree/master/tensorflow`
replace `tree/master` by `trunk`
and run following command in Terminal/CMD: 
`svn checkout https://github.com/tensorflow/tensorflow/trunk/tensorflow`
This will download tensorflow folder in pwd.



I hope this help. :)

**Comment by jdduke** [comment]
> Later on I'll want to do some pre- or post- processing to the inputs / outputs of the image - I assume that's where the C API differs from the C++ one - correct?

No, the C and C++ APIs are largely equivalent in terms of functionality. Note that you can still use C++ in your own library when using the C API. Also, the C API is self-contained, and has no absl/flatbuffer dependencies in the public API.

**Comment by orangesomethingorange** [comment]
> Hi @orangesomethingorange, we currently don't have any plans to expose the C++ API directly from the .aar. Generally speaking, using C++ across shared library boundaries is a tricky affair.
> 
> That said, we do expose the _C API_ for TFLite in the prebuilt .aars. We're working on a sample which downloads the aar from Maven and extracts them via Gradle. The .aar also include the C API headers. See also https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/c.

@jdduke I tried to build the C API library for Android like so:

### On my x86 Linux Ubuntu 16.04 [(following this):](https://www.tensorflow.org/lite/guide/android#build_tensorflow_lite_locally)

```
user@ub-16 $ bazel --version
bazel 3.1.0
user@ub-16 $ cat tensorflow/.tf_configure.bazelrc | grep -ie android
build --action_env ANDROID_NDK_HOME="/home/user/Android/Sdk/ndk/20.0.5594570/"
build --action_env ANDROID_NDK_API_LEVEL="28"
build --action_env ANDROID_BUILD_TOOLS_VERSION="29.0.3"
build --action_env ANDROID_SDK_API_LEVEL="29"
build --action_env ANDROID_SDK_HOME="/home/user/Android/Sdk"
user@ub-16 $ bazel build -c opt --cxxopt=--std=c++11 --config=android_arm64 //tensorflow/lite/c:tensorflowlite_c
...
ERROR: /home/user/.cache/bazel/_bazel_user/22c116ddaf0b061cb1d24c68b7eae373/external/local_config_cc/BUILD:47:1: in cc_toolchain_suite rule @local_config_cc//:toolchain: cc_toolchain_suite '@local_config_cc//:toolchain' does not contain a toolchain for cpu 'arm64-v8a'
ERROR: Analysis of target '//tensorflow/lite/c:tensorflowlite_c' failed; build aborted: Analysis of target '@local_config_cc//:toolchain' failed
INFO: Elapsed time: 0.209s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (9 packages loaded, 131 targets configured)

```


### On TensorFlow's Docker container [(following this):](https://www.tensorflow.org/lite/guide/build_arm64)

```
user@ub-16 $ sudo docker run -it --name tflite tensorflow/tensorflow:nightly-devel
root@148806b57557:~# apt-get update
...
Reading package lists... Done
root@148806b57557:~# apt-get install crossbuild-essential-arm64
...
root@148806b57557:/home# git clone https://github.com/tensorflow/tensorflow
root@148806b57557:/home/tensorflow# ./tensorflow/lite/tools/make/download_dependencies.sh
...
download_dependencies.sh completed successfully.
root@148806b57557:/home/tensorflow# bazel build -c opt --cxxopt=--std=c++11 --config=android_arm64 //tensorflow/lite/c:tensorflowlite_c
Extracting Bazel installation...
INFO: Options provided by the client:
  Inherited 'common' options: --isatty=1 --terminal_columns=101
INFO: Reading rc options for 'build' from /home/tensorflow/.bazelrc:
  Inherited 'common' options: --experimental_repo_remote_exec
WARNING: while reading option defaults file '/home/tensorflow/.bazelrc':
  invalid command name 'try-import'.
WARNING: while reading option defaults file '/home/tensorflow/.bazelrc':
  invalid command name 'try-import'.
ERROR: Unrecognized option: --experimental_repo_remote_exec
WARNING: --batch mode is deprecated. Please instead explicitly shut down your Bazel server using the command "bazel shutdown".
```

but it seems I do not have the proper toolchain for that.

Building with `bazel build -c opt //tensorflow/lite/c:tensorflowlite_c` build ok on x86 Ubuntu 16.04.
But when I try to include that in Android Studio like so:

**CMakeLists.txt**:
```
cmake_minimum_required(VERSION 3.4.1)
set(tflite_DIR /home/user/AndroidStudioProjects/tflite/app/libs )
add_library(
             native-lib
             SHARED
             native-lib.cpp )
add_library(libTensorflowLiteC
        SHARED
        IMPORTED)
set_target_properties(
        libTensorflowLiteC PROPERTIES IMPORTED_LOCATION
        ${tflite_DIR}/libTensorflowLiteC.so)
find_library(
              log-lib
              log )
target_link_libraries(
        native-lib
        libTensorflowLiteC
        ${log-lib} )
```

I get the following build error:

```
Build command failed.
Error while executing process /home/user/Android/Sdk/cmake/3.10.2.4988404/bin/ninja with arguments {-C /home/user/AndroidStudioProjects/tflite/app/.cxx/cmake/debug/arm64-v8a native-lib}
ninja: Entering directory `/home/user/AndroidStudioProjects/tflite/app/.cxx/cmake/debug/arm64-v8a'
[1/2] Building CXX object CMakeFiles/native-lib.dir/native-lib.cpp.o
[2/2] Linking CXX shared library /home/user/AndroidStudioProjects/tflite/app/build/intermediates/cmake/debug/obj/arm64-v8a/libnative-lib.so
FAILED: /home/user/AndroidStudioProjects/tflite/app/build/intermediates/cmake/debug/obj/arm64-v8a/libnative-lib.so 
: && /home/user/Android/Sdk/ndk/20.0.5594570/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=aarch64-none-linux-android28 --gcc-toolchain=/home/user/Android/Sdk/ndk/20.0.5594570/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/home/user/Android/Sdk/ndk/20.0.5594570/toolchains/llvm/prebuilt/linux-x86_64/sysroot -fPIC -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -fno-addrsig -Wa,--noexecstack -Wformat -Werror=format-security   -O0 -fno-limit-debug-info  -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libatomic.a -static-libstdc++ -Wl,--build-id -Wl,--warn-shared-textrel -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments -Wl,-z,noexecstack -shared -Wl,-soname,libnative-lib.so -o /home/user/AndroidStudioProjects/tflite/app/build/intermediates/cmake/debug/obj/arm64-v8a/libnative-lib.so CMakeFiles/native-lib.dir/native-lib.cpp.o  /home/user/AndroidStudioProjects/tflite/app/libs/libTensorflowLiteC.so /home/user/Android/Sdk/ndk/20.0.5594570/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/28/liblog.so -latomic -lm && :
**/home/user/AndroidStudioProjects/tflite/app/libs/libTensorflowLiteC.so: error adding symbols: File in wrong format**
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

How should I build the C API to be used in Android Studio?

**Comment by orangesomethingorange** [comment]
> @orangesomethingorange
> 
> Reply for steps you did in [comment](https://github.com/tensorflow/tensorflow/issues/39148#issuecomment-624075335)
> Whichever IDE you are using search there for the option to include additional dependencies. And then try adding path to TFlite folder. This will add your TFlite Libraries.
> For flatbuffers please read below.
> 
> I had same problem.
> Here's what I did.
> 
>     1. In my workspace, I created a folder `include` (for External dependencies eg. TFlite, flatbuffers, etc)
> 
>     2. Dowloaded only `tensorflow` folder from [source](https://github.com/tensorflow/tensorflow/tree/master/tensorflow). (You can do it using [svn](https://tortoisesvn.net/docs/release/TortoiseSVN_en/index.html))
> 
>     3. Similarly downloaded TFLite dependency [flatbuffers](https://github.com/google/flatbuffers/tree/master/include/flatbuffers) folder from the source.
> 
>     4. Also, TFlite needs absl header files so I downloaded [absl header files](https://github.com/abseil/abseil-cpp/tree/master/absl) from the source.
> 
>     5. I'm working in Microsoft VS. I added the `path/to/dir/include` in additional dependencies. Whichever IDE you are using search there for the option to include additional dependencies.
> 
> 
> Only the above folders are necessary no need to clone the complete repo.
> 
> After this include folder contains:
> include:
> --tensorflow
> ---- Some Folders (Including lite)
> ---- Some Files
> --flatbuffers
> ---- Some Folders
> ---- Some Files
> --absl
> ---- Some Folders
> ---- Some Files
> 
> Extra note:
> How to use svn. [stackoverflow post](https://stackoverflow.com/questions/7106012/download-a-single-folder-or-directory-from-a-github-repo).
> 
> I hope this help. :)

@pranv12,

My IDE is Android Studio 3.6.3.
To include **_flatbuffers_** and **_abseil_** I use their **_CMakeLists.txt_** files, trying to build them with Gradle, which fails with the following error:

`ninja: error: '../jniLibs/abseilLib/arm64-v8a/lib_abseil.a', needed by '/home/user/AndroidStudioProjects/tflite/app/build/intermediates/cmake/debug/obj/arm64-v8a/libnative-lib.so', missing and no known rule to make it`

**Comment by orangesomethingorange** [comment]
> (Not for Android) on Ubuntu 16.04:

@jdduke - continuing ### (Not for Android) on Ubuntu 16.04:

I created a fresh Android Studio C++ empty project and used the following **CMakeLists.txt**:
```
cmake_minimum_required(VERSION 3.4.1)
set( lib_tflite_DIR ../jniLibs )
file(MAKE_DIRECTORY ${lib_tflite_DIR})
file(MAKE_DIRECTORY ${lib_tflite_DIR}/${ANDROID_ABI})
add_library( tflite-lib
        SHARED
        IMPORTED )
set_target_properties(
        tflite-lib
        PROPERTIES IMPORTED_LOCATION
        app/src/main/jniLibs/x86_64/libtensorflowlite_c.so )
add_library(
             native-lib
             SHARED
             native-lib.cpp )
find_library(
              log-lib
              log )
target_link_libraries(
        native-lib
        tflite-lib
        ${log-lib} )
```

and got the following error trying to build:

```
Build command failed.
Error while executing process /home/user/Android/Sdk/cmake/3.10.2.4988404/bin/ninja with arguments {-C /home/user/AndroidStudioProjects/tflite/app/.cxx/cmake/debug/arm64-v8a native-lib}
ninja: Entering directory `/home/user/AndroidStudioProjects/tflite/app/.cxx/cmake/debug/arm64-v8a'

**ninja: error:** 'app/src/main/jniLibs/x86_64/libtensorflowlite_c.so', **needed by** '/home/user/AndroidStudioProjects/tflite/app/build/intermediates/cmake/debug/obj/arm64-v8a/libnative-lib.so', **missing and no known rule to make it**
```

Trying to change `BUILD` target in bazel or create a new target did not prove successful either.
Is buildting the C API libs from Gradle's `CMakeLists.txt` an option?

**Comment by jdduke** [comment]
In your CMakeLists.txt it looks like you've specified the x86_64 binary, whereas  the build error is looking in the arm64 directory?

`ninja: Entering directory /home/user/AndroidStudioProjects/tflite/app/.cxx/cmake/debug/arm64-v8a
`

**Comment by orangesomethingorange** [comment]
I have the same `.so` library file placed in both directories, with architecture-names accordingly, just in case Gradle (or ninja) wants to look for them there (i.e. `app/src/main/jniLibs/arm64-v8a/[LIB].so` and `app/src/main/jniLibs/x86_64/[LIB].so`).

In my `CMakeLists.txt` file I have:
```
set_target_properties(
        tflite-lib
        PROPERTIES IMPORTED_LOCATION
        **app/src/main/jniLibs/x86_64/libtensorflowlite_c.so** )
```
so I guess that's why the build error looks there.

I specified both architectures at build.
Changing the line to `app/src/main/jniLibs/arm64-v8a/libtensorflowlite_c.so` results in the same error, just with this path:
`ninja: error: 'app/src/main/jniLibs/arm64-v8a/libtensorflowlite_c.so', needed by '/home/user/AndroidStudioProjects/tflite/app/build/intermediates/cmake/debug/obj/arm64-v8a/libnative-lib.so', missing and no known rule to make it`

Also, taking the following line as an example:
`bazel build -c opt
--fat_apk_cpu=x86_64,arm64-v8a --cpu=x86_64,arm64-v8a
--host_crosstool_top=@bazel_tools//tools/cpp:toolchain
//tensorflow/lite/java:tensorflow-lite`

I tried various architecture combinations:
```
--fat_apk_cpu=x86_64,arm64-v8a --cpu=x86_64,arm64-v8a

--fat_apk_cpu=x86_64 --cpu=x86_64

--fat_apk_cpu=aarch64 --cpu=aarch64

--cpu=x86_64,arm64-v8a

--fat_apk_cpu=x86_64,arm64-v8a

--fat_apk_cpu=arm64-v8a --cpu=x86_64
```
but build fails.

This line use the CPP toolchain on the `tensorflow/lite/java` library, following [this example](https://www.tensorflow.org/lite/guide/android#build_and_install), but I fail to find the toolchain code in the bazel BUILD files.

Could this be the issue? I'm only just getting to know bazel now...

**Comment by orangesomethingorange** [comment]
> > Later on I'll want to do some pre- or post- processing to the inputs / outputs of the image - I assume that's where the C API differs from the C++ one - correct?
> 
> No, the C and C++ APIs are largely equivalent in terms of functionality. Note that you can still use C++ in your own library when using the C API. Also, the C API is self-contained, and has no absl/flatbuffer dependencies in the public API.

@jdduke I think I misunderstood your answer. I thought that having the C API being self-contained and free from absl/flatbuffer dependencies means I can just use the files.
After failing to build the library as a shared / static library to include in my project, I tried to build and compile the source and header files directly in the IDE. I started from the example in the `c_api.h` file, and added headers one by one, until I got to here:
```
$ tree
.
|-- absl
|   |-- base
|   |   |-- attributes.h
|   |   |-- config.h
|   |   |-- internal
|   |   |   |-- identity.h
|   |   |   |-- inline_variable.h
|   |   |   `-- invoke.h
|   |   |-- macros.h
|   |   |-- optimization.h
|   |   |-- options.h
|   |   |-- policy_checks.h
|   |   `-- port.h
|   |-- memory
|   |   `-- memory.h
|   |-- meta
|   |   `-- type_traits.h
|   |-- types
|   |   |-- bad_optional_access.h
|   |   |-- internal
|   |   |   `-- optional.h
|   |   `-- optional.h
|   `-- utility
|       `-- utility.h
|-- flatbuffers
|   |-- base.h
|   |-- flatbuffers.h
|   `-- stl_emulation.h
`-- tensorflow
    |-- core
    |   `-- public
    |       `-- version.h
    `-- lite
        |-- allocation.h
        |-- c
        |   |-- c_api.h
        |   |-- c_api_internal.h
        |   `-- common.h
        |-- core
        |   |-- api
        |   |   |-- error_reporter.h
        |   |   |-- op_resolver.h
        |   |   `-- profiler.h
        |   |-- macros.h
        |   `-- subgraph.h
        |-- delegates
        |   `-- nnapi
        |       `-- nnapi_delegate.h
        |-- error_reporter.h
        |-- experimental
        |   `-- resource
        |       `-- resource_base.h
        |-- external_cpu_backend_context.h
        |-- interpreter.h
        |-- interpreter_builder.h
        |-- kernels
        |   `-- register.h
        |-- memory_planner.h
        |-- model.h
        |-- model_builder.h
        |-- mutable_op_resolver.h
        |-- nnapi
        |   |-- NeuralNetworksTypes.h
        |   `-- nnapi_implementation.h
        |-- schema
        |   `-- schema_generated.h
        |-- stderr_reporter.h
        |-- string_type.h
        |-- type_to_tflitetype.h
        |-- util.h
        `-- version.h
```
this was the minimal structure I got to by following header files.
As you can see - absl/flatbuffer dependencies do exist.

### Issue 1
Trying to build from this structure results in build errors of the form:
`error: undefined reference to '[FUNCTION]'` which all relate to functions found in `c_api.cc`.

### Issue 2
Trying to add and build with `c_api.cc` included,

results in the following error from `<memory>`:
`.../toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/memory:2339: error: undefined reference to 'tflite::impl::Interpreter::~Interpreter()'`
when building for `x86_64`

and the following errors:
```
c_api.cc:62: undefined reference to `tflite::DefaultErrorReporter()'
c_api.cc:62: undefined reference to `tflite::FlatBufferModel::VerifyAndBuildFromBuffer(char const*, unsigned long, tflite::TfLiteVerifier*, tflite::ErrorReporter*)'
[[ many more "undefined reference to" from c_api.cc ]]
...
mutable_op_resolver.h:59: undefined reference to `vtable for tflite::MutableOpResolver'
[[ likewise ]]
...
memory:2339: undefined reference to `tflite::FlatBufferModel::~FlatBufferModel()'
c_api.cc.o: In function `std::__ndk1::default_delete<tflite::impl::Interpreter>::operator()(tflite::impl::Interpreter*) const':
```
when trying to build for `arm64-v8a`.

This was done using clang++ in the toolchain.

How can I use the C API independently?

**Comment by jdduke** [comment]
>  I thought that having the C API being self-contained and free from absl/flatbuffer dependencies means I can just use the files.

Sorry if it wasn't clear, what I meant is that the C API *headers* are free from external dependencies. So you can easily drop the shared library + 2-3 lite/c/*.h headers into your project and use them directly, as opposed to also needing flatbuffer headers. The actual C API *implementation* requires all the same libraries as the C++ API.

**Comment by orangesomethingorange** [comment]
> > I thought that having the C API being self-contained and free from absl/flatbuffer dependencies means I can just use the files.
> 
> Sorry if it wasn't clear, what I meant is that the C API _headers_ are free from external dependencies. So you can easily drop the shared library + 2-3 lite/c/*.h headers into your project and use them directly, as opposed to also needing flatbuffer headers. The actual C API _implementation_ requires all the same libraries as the C++ API.

OK, that's great news but how do I use them?
The only example of running inference with the C-API is the one in the `c_api.h` ([this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/c/c_api.h)), right?

Do you have an example code for using only the C headers and still running an inference?

**Comment by jdduke** [comment]
You're right that there's no good end-to-end example using the C API. This is something we're working on, however, the high-level usage and flow should be quite similar to the C++ API.

**Comment by orangesomethingorange** [comment]
Well that's good to hear :) I hope to see them soon.
Currently I'm having trouble with just using the C/C++ APIs.

The main approaches I've been trying so far are:
1. using the TFLite as an external library and load into Android Studio: I used [this](https://www.tensorflow.org/lite/guide/android#build_tensorflow_lite_locally), and [this](https://www.tensorflow.org/lite/guide/build_arm64) and combinations of the two, to get a static (.a) or a shared (.so) library (from either my x86-toolchain or the Docker container), and try to include them in Android Studio using cmake.
2. building the source code from files + headers with Gradle (Android Studio's build system): I used the C source code from lite's directory ([see here](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite)) and tried to build it in Android Studio.

Any comments / ideas are more than welcome.

**Comment by Saduf2019** [comment]
@orangesomethingorange 
Is this still an issue, please confirm if we may move this to closed status

**Comment by orangesomethingorange** [comment]
> @orangesomethingorange
> Is this still an issue, please confirm if we may move this to closed status

Hey @Saduf2019, I can't say that I got an answer for this issue.
Also, I'm still working on investigating this, and I understand from @jdduke that this might be solved in the near future, since work is currently (or very soon will be) in progress.

I would like to keep the status as open for now, if that's OK with you :)

**Comment by orangesomethingorange** [comment]
I recreated the shared libraries, and compiled with `cmake` normally and now it works OK.

In my case the issue was the shared libraries build, which wasn't updated in container versions.
Generally speaking the Docker Containers simplify the build of shared libraries, but it really depends on the container.
For example - using the [latest, devel, latest-devel or nightly containers](https://hub.docker.com/r/tensorflow/tensorflow/tags/) made a big difference, and I'm still not sure which one contains what exactly and how configurable each one is.
Also note the [Docker Container from the Dockerfile](https://raw.githubusercontent.com/tensorflow/tensorflow/master/tensorflow/tools/dockerfiles/tflite-android.Dockerfile).

Once I found proper containers for each shared library I could build them with customization (for example [ops](https://www.tensorflow.org/lite/guide/ops_custom) or [custom ops](https://www.tensorflow.org/lite/guide/ops_custom)).

After that the process it's pretty much the same as I described at the beginning of this issue, and it's a great replacement to the use of AARs in the Gradle project.
